### ISSUE TITLE
invalid userdata due to optional attributes

### ISSUE THREAD

**Issue by lobeck** [terraform-aws-modules/terraform-aws-eks#2787]
## Description
- [X] ✋ I have searched the open/closed issues and my issue is not listed.

TLDR: The variables `cluster_endpoint` and `cluster_auth_base64` are optional, this allows for a semi broken userdata.

We are configuring the EKS cluster and nodegroups separate, so we manually call eks-managed-node-group. 

We have now noticed, that we're not passing the endpoint and certificate values, as those are optional. However the `_user_data`
 / template don't cope with that properly.

This results in a userdata with empty variables:

    B64_CLUSTER_CA=
    API_SERVER_URL=
    /etc/eks/bootstrap.sh go-main-prod  --b64-cluster-ca $B64_CLUSTER_CA --apiserver-endpoint $API_SERVER_URL
    echo "go get 'em little kubelet!"

During boot we thus get:

    cloud-init[3329]: + /etc/eks/bootstrap.sh go-main-prod --b64-cluster-ca --apiserver-endpoint
    cloud-init[3329]: 2023-10-18T14:39:58+0000 [eks-bootstrap] INFO: starting...
    cloud-init[3329]: 2023-10-18T14:39:58+0000 [eks-bootstrap] INFO: --b64-cluster-ca='--apiserver-endpoint'
    ...
    cloud-init[3329]: 2023-10-18T14:39:58+0000 [eks-bootstrap] INFO: --cluster-ca or --api-server-endpoint is not defined, describing cluster...

So only because of the order, the bootstrap is not failing completely.

## Versions

- Module version [Required]: v.19.17.2

- Terraform version: not relevant
- Provider version(s): not relevant

## Reproduction Code [Required]

calling eks_worker_groups without `cluster_endpoint` and `cluster_auth_base64`

    module "eks_worker_groups" {
      source  = "terraform-aws-modules/eks/aws//modules/eks-managed-node-group"
      version = "v19.15.4"
      ....

not relevant

## Expected behavior

`cluster_endpoint` and `cluster_auth_base64` should be mandatory
OR
`_user_data` should filter

## Actual behavior

semi-broken userdata gets created

**Comment by bryantbiggs** [comment]
> cluster_endpoint and cluster_auth_base64 should be mandatory

They are mandatory, otherwise nodes do not join the cluster

**Comment by lobeck** [comment]
Well, fortunately the bootstrap.sh has it's internal protection against that. That's why nobody noticed it yet.

**Comment by bryantbiggs** [comment]
apologies, `cluster_name` is required - the rest are optional, in theory, but will incur an describe API call during provisioning

Is there an issue you are facing?

**Comment by lobeck** [comment]
You're right. Just noticed, that these flags are indeed optional. But then the template should also treat them as such. The current code generation feels dangerous.

https://github.com/awslabs/amazon-eks-ami/blob/master/files/bootstrap.sh#L23-L25 describes them as optional.

I'm not having issues and if you don't mind, I'll update _user_data to not pass them when empty

**Comment by bryantbiggs** [comment]
for this module I think its better to stick with what is provided where the values known from the control plane are passed into the bootstrap userdata. This reduces the number of describe API calls during instance provisioning. Closing this for now but please feel free to respond if I have missed something
